### PR TITLE
Rename input params for community and direct shared galleries & add validations

### LINKED ISSUE #203: Rename communityGallery_image_id and directSharedGallery_image_id to follow snake case
#### Overview of the Issue

Rename communityGallery_image_id and directSharedGallery_image_id to follow snake case.
Also add checks for mutal exclusiveness

#### Reproduction Steps

try to provide multiple kinds of gallery as input

### PR DESCRIPTION
**Pull request by agrawalkshitij**
Includes the following changes:
1) Renames community gallery and direct shared gallery image ids to follow snake naming
2) Add extra validation checks to make sure only one kind of gallery can be provided as input
3) Added some unit tests
4) Fixed some very old unit tests which didn't seem right. [extra]

Also clarified in comments why the gallery name and gallery unique names are 2 diff things, and thus the same fields should not be used.

Closes #203

### CODE REVIEW

**Comment by agrawalkshitij** [comment]
@JenGoldstrich - pls take a look whenever you get time! This hopefully fixes any issues mentioned before!

**Review by invidian — APPROVED** [review]
LGTM, thanks for the quick follow up!

File: builder/azure/arm/config.go [hunk 1]
```diff
@@ -78,7 +78,7 @@ type PlanInformation struct {
 type SharedImageGallery struct {
 	Subscription  string `mapstructure:"subscription"`
 	ResourceGroup string `mapstructure:"resource_group"`
-	GalleryName   string `mapstructure:"gallery_name"`
+	GalleryName   string `mapstructure:"gallery_name"` //Gallery resource name
```
[the remark this answers is not in the document]
Comment (agrawalkshitij) [reply builder/azure/arm/config.go:81]: I wanted to explain the difference b/w gallery name and gallery unique name

File: builder/azure/arm/config.go [hunk 2]
```diff
@@ -140,8 +140,8 @@ type Config struct {
 	//     "gallery_name": "GalleryName",
 	//     "image_name": "ImageName",
 	//     "image_version": "1.0.0",
-	//     "communityGallery_image_id": "/CommunityGalleries/{cg}/Images/{}/Versions/{}",
-	//     "directSharedGallery_image_id": "/SharedGalleries/{cg}/Images/{}/Versions/{}"
+	//     "community_gallery_image_id": "/CommunityGalleries/{galleryUniqueName}/Images/{}/Versions/{}",
+	//     "directSharedGallery_image_id": "/SharedGalleries/{galleryUniqueName}/Images/{}/Versions/{}"
```
[the remark this answers is not in the document]
Comment (agrawalkshitij) [reply builder/azure/arm/config.go]: fixed! thanks for catching!

**Comment by agrawalkshitij** [comment]
@JenGoldstrich @nywilken - can you help take a look?

File: builder/azure/arm/config.go [hunk 3]
```diff
@@ -87,11 +87,11 @@ type SharedImageGallery struct {
...
 	ImageVersion string `mapstructure:"image_version" required:"false"`
 
-	// Id of the community gallery image : /CommunityGalleries/{cgName}/Images/{img}[/Versions/{}] (Versions part is optional)
-	CommunityGalleryImageId string `mapstructure:"communityGallery_image_id" required:"false"`
+	// Id of the community gallery image : /CommunityGalleries/{galleryUniqueName}/Images/{img}[/Versions/{}] (Versions part is optional)
+	CommunityGalleryImageId string `mapstructure:"community_gallery_image_id" required:"false"`
 
-	// Id of the direct shared gallery image : /sharedGalleries/{cgName}/Images/{img}[/Versions/{}] (Versions part is optional)
-	SharedGalleryImageID string `mapstructure:"directSharedGallery_image_id" required:"false"`
+	// Id of the direct shared gallery image : /sharedGalleries/{galleryUniqueName}/Images/{img}[/Versions/{}] (Versions part is optional)
```
Comment (nywilken) [inline builder/azure/arm/config.go:93]: If an image version is provided in the community or direct shared gallery image does the value of ImageVersion get used? 

Do they both need to be identical in order for things to work?

**Comment by nywilken** [comment]
I updated the documentation to highlight the new Shared Image Gallery block configs. In the last commit I replace privately with private. 

image

File: builder/azure/arm/config.go (hunk 3 above)
Comment (nywilken) [reply builder/azure/arm/config.go:93]: I went ahead and updated the documentation to include these new configuration. I think there are a few inconsistencies that we should fast follow on. But I did want to get the documentation in a better place for now.

**Review by JenGoldstrich — APPROVED** [review]
Thanks for your review and docs changes @nywilken!

**Comment by agrawalkshitij** [comment]
Hey,

 I am on vacation on right now, so not able to push new iterations with comment fixes. Did you find everything OK, last I checked there were some comments which needed some changes. Did one of you pick up those changes? If yes, thanks a lot. If no, I will get to it once I get back.

Thanks,
Kshitij

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by nywilken** [comment]
> Hey, I am on vacation on right now, so not able to push new iterations with comment fixes. Did you find everything OK, last I checked there were some comments which needed some changes. Did one of you pick up those changes? If yes, thanks a lot. If no, I will get to it once I get back. Thanks, Kshitij Get Outlook for iOS<https://aka.ms/o0ukef>
> […](#)
> ________________________________ From: Jenna Goldstrich ***@***.***> Sent: Tuesday, June 28, 2022 9:47:41 PM To: hashicorp/packer-plugin-azure ***@***.***> Cc: Kshitij Agrawal ***@***.***>; Mention ***@***.***> Subject: Re: [hashicorp/packer-plugin-azure] Rename input params for community and direct shared galleries & add validations (PR #204) Merged #204<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fhashicorp%2Fpacker-plugin-azure%2Fpull%2F204&data=05%7C01%7Ckagarwal%40microsoft.com%7Cbe0bdfe763254499981d08da5921ba64%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637920298646525583%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=plDII1z9UA9gKZ5eZY%2FVknhiECIB%2FdNqNaADshsaDRc%3D&reserved=0> into main. — Reply to this email directly, view it on GitHub<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fhashicorp%2Fpacker-plugin-azure%2Fpull%2F204%23event-6895503957&data=05%7C01%7Ckagarwal%40microsoft.com%7Cbe0bdfe763254499981d08da5921ba64%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637920298646525583%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=7uR2PDKFyaqTAP6CmjJstzMwKTgO3aUSEscSMm2EW0I%3D&reserved=0>, or unsubscribe<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAMEH2BYOIP7EZBYQSV6ZTG3VRMQSLANCNFSM5XXCFNAA&data=05%7C01%7Ckagarwal%40microsoft.com%7Cbe0bdfe763254499981d08da5921ba64%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637920298646525583%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=lQDda8jNhCl7ncOogBILQFnBw9RqSzuxJmAfCznje3g%3D&reserved=0>. You are receiving this because you were mentioned.Message ID: ***@***.***>

@agrawalkshitij we pushed a few changes to the comments to address the changes I requested. There are still some open questions around the usage of the galleries, which we can follow up when back. We are going to release this to address the invalid configuration argument names. We will bump the minor version to denote a breaking change.

**Pull request merged**